### PR TITLE
Add UMLS API

### PR DESCRIPTION
**Pull request by mirestrepo**
Interface to the Unified Medical Language System API

- Add basic API methods
- Add related documentation
- Add related tests

### CODE REVIEW

**Comment by mirestrepo** [comment]
I have set up tests - however they need Environment variables in TRAVIS... I have set those up as well but they are not available from PR. Therefore the decrease in coverage and

**Comment by bicycle1885** [comment]
Can you write a quick and executable manual so that people not familiar with the service can try the APIs? I mean something like this: https://github.com/BioJulia/BioServices.jl/blob/ad7d4eb8c539a5192e90216039bd85c5004fea2a/docs/src/man/eutils.md.

**Comment by mirestrepo** [comment]
I've added more details and a sample workflow to the docs to make it more straight forward for those not familiar with the service. Let me know what you think.

File: docs/src/man/umls.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    ```
+    using BioServices.UMLS
+    ```
+
+## Available Endpoints
+
+For a  complete list of the enpoints made available in the REST API visit the [REST API Documentation](https://documentation.uts.nlm.nih.gov/rest/home.html)
+
+This module focuses in the following three enpoints. (Request to expand API is encouraged trhough pull requests or issues in out GitHub repository.)
```
Comment (bicycle1885) [inline docs/src/man/umls.md]: - trhough -> through?
- out -> our?

File: docs/src/man/umls.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+| [`get_semantic_types`](#semantic-types) | Retrieve smenatic types associated with a CUI|
+
+-------------------
+
+## Sample workflow
+
+Service tickets are needed each time you search or retrieve content from the UMLS REST API. 
+A service ticket is retrieved automatically by this software from a **ticket granting ticket**.
+
+Therefore, the firts step of your workflow must start by requesting a **ticket granting ticket**
```
Comment (bicycle1885) [inline docs/src/man/umls.md]: - firts -> first?

File: docs/src/man/umls.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+* Exact Match: retrieves only concepts that include a synonym that exactly matches the search term.
+* Normalized String: use with English language terms only. Removes lexical variations such as plural and upper case text and compares search terms to the Metathesaurus normalized string index to retrieve relevant concepts.
+* Normalized Word: use with English language terms only. Removes lexical variations such as plural and upper case text, and compares search terms to the Metathesaurus normalized word index to retrieve relevant concepts.
+* Right Truncation: retrieves concepts with synonyms that begin with the letters of the search term. For example, a right truncation search for "bronch" retrieves concepts that contain synonyms such as bronchitis, bronchiole, bronchial artery.
+* Left Truncation: retrieves concepts with synonyms that end with the letters of the search term. For example, a left truncation search for "itis" retrieves concepts that contain synonyms such as colitis, bronchitis, pancreatitis.
+
+--------------------------------------------------
```
Comment (bicycle1885) [inline docs/src/man/umls.md]: I think these horizontal lines are not needed.

File: src/umls.jl [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+# License is MIT: https://github.com/BioJulia/Bio.jl/blob/master/LICENSE.md
+
+"""
+Julia interface to [Unified Medical Language REST API](https://github.com/HHS/uts-rest-api)
+
+
+The UMLS provides basic functionality to authenticate and query the UMLS API.
+The following functions are exported from this module:
+1. `get_tgt`: Get Ticket-Granting Ticket
+3. `search_umls`: Search the UMLS
```
Comment (bicycle1885) [inline src/umls.jl]: No.2 is missing.

File: src/umls.jl [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+#-------------- Authentication -------------------
+"""
+time_to_last_save(file)
+
+Get how many hours since last save 
+"""
+function time_to_last_save(file)
+#unix time is GMT
+time_diff = Dates.value(Dates.unix2datetime(time()) - Dates.unix2datetime(mtime(file)))/ (1000 * 60 * 60)
+return time_diff
```
Comment (bicycle1885) [inline src/umls.jl]: Indent code between `function` and `end`.

File: src/umls.jl [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+#-------------- Authentication -------------------
+"""
+time_to_last_save(file)
+
+Get how many hours since last save 
+"""
+function time_to_last_save(file)
+#unix time is GMT
+time_diff = Dates.value(Dates.unix2datetime(time()) - Dates.unix2datetime(mtime(file)))/ (1000 * 60 * 60)
```
Comment (bicycle1885) [inline src/umls.jl]: I think you can use `now()` instead  `Dates.unix2datetime(time())`.

File: src/umls.jl [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    code::Int64
+end
+
+function Base.showerror(io::IO, e::BadResponseException)
+    print(io, "BadResponseException with code: ",STATUS_CODES[e.code])
+end
+
+#-------------- Authentication -------------------
+"""
+time_to_last_save(file)
```
Comment (bicycle1885) [inline src/umls.jl]: Indent the first line of a docstring.

File: src/umls.jl [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+    try
+        ticket = getattr(doc.root.children[2].children[2], "action")
+
+        open(tgt_file, "w") do f
+            write(f, ticket)
+        end
+
+        return ticket
+    catch
+        error("Could not get TGT: UTS response structure is wrong")
```
Comment (bicycle1885) [inline src/umls.jl]: An exception may happen when opening and writing a ticket to a file in the try block. It should be distinguished from the invalid response.

File: src/umls.jl [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+"""
+get_ticket(tgt)
+
+Retrieve a single-use Service Ticket using TGT
+"""
+function get_ticket(tgt)
+    params = Dict("service"=> service)
+    h = Dict("Content-type"=> "application/x-www-form-urlencoded",
+    "Accept"=> "text/plain", "User-Agent"=>"bcbi-julia" )
```
Comment (bicycle1885) [inline src/umls.jl]: Why "bcbi-julia"? Just "julia" or "BioServices.jl" may be more meaningful.

File: src/umls.jl [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    result_pages = Array{Any,1}()
+
+    while true
+
+        #get a new ticket per page if necessary
+        ticket = ""
+        try
+            ticket = get_ticket(tgt)
+        catch err
+            rethrow(err)
```
Comment (bicycle1885) [inline src/umls.jl]: I think you don't need to catch and rethrow an exception here.

File: src/umls.jl [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        end
+
+        page +=1
+        #append ticket to query
+        query["ticket"]= ticket
+        query["pageNumber"]= @sprintf("%d", page)
```
Comment (bicycle1885) [inline src/umls.jl]: You can use `string(page)` to stringify an integer.

File: src/umls.jl [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+sm = get_semantic_types(tgt, cui)
+```
+"""
+function get_semantic_types(tgt, cui; version="current")
+    json_response = get_cui(tgt,cui; version=version)
+    st = json_response["result"]["semanticTypes"]
+    concepts = Array{String}(length(st))
+    for (ci, concept) in enumerate(st)
+        concepts[ci] =concept["name"]
+    end
```
Comment (bicycle1885) [inline src/umls.jl]: `return [String(concent["name"]) for concept in st]` may be easier to read.

File: src/umls.jl [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+
+    return concepts
+end
+
+end
```
Comment (bicycle1885) [inline src/umls.jl]: Append a newline at the end.

**Comment by mirestrepo** [comment]
@bicycle1885 Thank you for all the insightful comments. I'm merging, verifying all tests run with the environment variable set in Travis and I'll tag and do the first release

**Comment by Ward9250** [comment]
Great stuff! We have attobot monitoring BioServices so tagging the release should be all that is necessary for this to get put on Metadata.jl

**Pull request merged**